File: backend/app/models/replenishment.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class MLModelRun(Base):
    __tablename__ = "ml_model_runs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    model_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    model_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        default="RandomForestRegressor",
    )

    target_column: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        default="sales",
    )

    training_row_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    feature_columns: Mapped[list[str]] = mapped_column(
        JSONB,
        nullable=False,
    )

    mae: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    rmse: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    r2: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    model_path: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
    )

    trained_at: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    recommendations = relationship(
        "ReplenishmentRecommendation",
        back_populates="model_run",
    )


class ReplenishmentRecommendation(Base):
    __tablename__ = "replenishment_recommendations"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    model_run_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("ml_model_runs.id"),
        nullable=True,
        index=True,
    )

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("products.id"),
        nullable=False,
        index=True,
    )

    sku: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        index=True,
    )

    product_name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )

    category: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    material: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    color: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    location: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    store_type: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    season: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    price: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    cost: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    current_inventory: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
    )

    available_inventory: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
    )

    incoming_inventory: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
    )

    inventory_supply: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
    )

    predicted_period_sales: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    predicted_daily_sales: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    forecasted_demand_during_lead_time: Mapped[
        Decimal | None
    ] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    average_delivery_days: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    safety_stock: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    reorder_point: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    reorder_gap: Mapped[Decimal | None] = mapped_column(
        Numeric(12, 2),
        nullable=True,
    )

    predicted_inventory_sales_ratio: Mapped[
        Decimal | None
    ] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    suggested_reorder_quantity: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
    )

    replenishment_priority_score: Mapped[
        Decimal | None
    ] = mapped_column(
        Numeric(12, 4),
        nullable=True,
    )

    replenishment_decision: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    risk_level: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )

    reason: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default="pending",
        index=True,
    )

    generated_at: Mapped[datetime] = mapped_column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
    )

    reviewed_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    model_run = relationship(
        "MLModelRun",
        back_populates="recommendations",
    )

    product = relationship("Product")

File: frontend/src/types/replenishment.ts
export interface ActiveModelStatus {
  available: boolean;

  model_run_id: string | null;
  model_name: string | null;
  model_type: string | null;
  target_column: string | null;

  training_row_count: number | null;

  mae: string | null;
  rmse: string | null;
  r2: string | null;

  is_active: boolean | null;
  trained_at: string | null;
}

export interface DemandModelTrainingRequest {
  data_source: string | null;
}

export interface FeatureImportance {
  feature: string;
  importance: number;
}

export interface DemandModelTrainingResponse {
  model_run_id: string;
  model_name: string;
  model_type: string;
  target_column: string;

  training_row_count: number;
  training_set_row_count: number;
  test_set_row_count: number;

  mae: number;
  rmse: number;
  r2: number;

  model_path: string;
  trained_at: string;

  feature_columns: string[];
  categorical_features: string[];
  numeric_features: string[];

  top_feature_importances: FeatureImportance[];
}

export interface ReplenishmentGenerationRequest {
  demand_period_days: number;
  lead_time_days: number;
  safety_stock_days: number;
  minimum_order_quantity: number;

  include_in_production_as_incoming: boolean;
  include_in_transit_as_incoming: boolean;

  urgent_gap_threshold: number;
  high_priority_threshold: number;
  medium_priority_threshold: number;
}

export interface ReplenishmentGenerationResponse {
  model_run_id: string;

  generated_count: number;
  high_risk_count: number;
  medium_risk_count: number;
  low_risk_count: number;

  total_suggested_quantity: number;
  generated_at: string;
}

export type ReplenishmentRiskLevel =
  | "High"
  | "Medium"
  | "Low";

export type ReplenishmentDecision =
  | "Urgent Replenishment"
  | "Consider Replenishment"
  | "Maintain Current Inventory"
  | "Slow Down Replenishment"
  | "Reduce Future Purchasing";

export interface ReplenishmentRecommendation {
  id: string;
  model_run_id: string | null;
  product_id: string;

  sku: string;
  product_name: string;

  category: string | null;
  material: string | null;
  color: string | null;
  location: string | null;
  store_type: string | null;
  season: string | null;

  price: string | null;
  cost: string | null;

  current_inventory: number;
  available_inventory: number;
  incoming_inventory: number;
  inventory_supply: number;

  predicted_period_sales: string | null;
  predicted_daily_sales: string | null;
  forecasted_demand_during_lead_time: string | null;

  average_delivery_days: string | null;
  safety_stock: string | null;
  reorder_point: string | null;
  reorder_gap: string | null;

  predicted_inventory_sales_ratio: string | null;

  suggested_reorder_quantity: number;
  replenishment_priority_score: string | null;

  replenishment_decision: ReplenishmentDecision;
  risk_level: ReplenishmentRiskLevel;
  reason: string | null;

  status: string;
  generated_at: string;
  reviewed_at: string | null;
}